lib/cdo/test_flakiness.rb

```diff
@@ -76,23 +70,52 @@ def self.recommend_reruns(flakiness)
     return [max_reruns, confidence]
   end
 
-  CACHE_FILENAME = (File.dirname(__FILE__) + "/../../dashboard/tmp/cache/flakiness.json").freeze
+  CACHE_FILENAME = (File.dirname(__FILE__) + "/../../dashboard/tmp/cache/test_summary.json").freeze
   CACHE_TTL = 86400 # 1 day of seconds
 
-  def self.cache_test_flakiness
+  using CacheMethod
+
+  cached def self.test_summary
     if File.exist?(CACHE_FILENAME) &&
         (Time.now - File.mtime(CACHE_FILENAME)) < CACHE_TTL
       return JSON.parse(File.read(CACHE_FILENAME))
     end
 
-    @@test_flakiness = calculate_test_flakiness
+    summarize_by_job.reject {|_, s| s['total'] < MIN_SAMPLES}.tap do |summary|
+      File.write(CACHE_FILENAME, JSON.dump(summary))
+    end
+  end
 
-    File.open(CACHE_FILENAME, 'w') {|f| f.write(JSON.dump(@@test_flakiness))}
+  cached def self.test_flakiness
+    test_summary.transform_values {|s| s['fail_rate']}
+  end
+
+  cached def self.test_duration
+    test_summary.transform_values {|s| s['duration'].round(2)}
+  end
 
-    @@test_flakiness
+  cached def self.test_estimate
+    test_summary.transform_values {|s| (recommend_reruns(s['fail_rate']).first * s['duration']).round(2)}
```

Comment:
This is very cool, and I'd love to see this used as our default in the future.

I do have a few questions:

- You're using `max_reruns * avg_duration`. I was wondering if we should be using `(max_reruns / 2) * avg_duration`, assuming a flaky test will trend toward using half of its max_reruns on average?
- The computed average duration only takes successful runs into account, but it's common for passing runs and failing runs to have very different timing patterns.  I wonder if it's worth computing average failure duration from the samples we've pulled and changing this calculation to something closer to `(expected_runs - 1) * avg_failure_duration + avg_success_duration`
  e.g. A test that normally passes in 30 seconds but often fails with a 5-minute timeout with 4 max_reruns would currently get a 2-minute estimate, but its worst (passing) case is actually 20:30.